### DISCUSSION TITLE
AlphaFold2 error

### DISCUSSION THREAD

**Discussion by Shunsuke-Kanamori** [nig-sc/nigsc_homepage2#473] [Q&A] [answered]
こちらのサイト（https://sc.ddbj.nig.ac.jp/advanced_guides/Alphafold_2_2 ） を参考に、AlphaFold2を実行してみたのですが、以下のエラーが出てしまいます。何か解決策のわかる方は教えていただけると幸いです。よろしくお願いします。

実行したスクリプト
```
#!/bin/sh
#$ -S /bin/sh
#$ -cwd
ulimit -s unlimited

#$ -l s_vmem=20G
#$ -l mem_req=20G
#$ -pe def_slot 16

echo start at
date



FASTAFILE=input.fa
OUTPUTDIR=out
DATE="2021-11-12"
MODEL="monomer"
PRED=5

export OPENMM_CPU_THREADS=16
export XLA_FLAGS="--xla_cpu_multi_thread_eigen=false intra_op_parallelism_threads=16"

singularity exec \
-B /lustre7/software/alphafold/database:/lustre7/software/alphafold/database \
-B /lustre7/software/alphafold/2.2.2/database:/data1/database \
/lustre7/software/alphafold/2.2.2/alphafold-2.2.2-CPU.sif \
/opt/alphafold/bin/alphafold \
--fasta_paths=${FASTAFILE} \
--output_dir=${OUTPUTDIR} \
--model_preset=${MODEL} \
--max_template_date=${DATE} \
--use_gpu_relax=false \
--num_multimer_predictions_per_model=${PRED}



echo ending at
date
```

エラーメッセージ

> /home/kanashun/.profile: line 1: /lustre7/home/lustre3/kanashun/homebrew/bin/brew: No such file or directory
> I0418 15:18:44.865874 46912496321088 templates.py:857] Using precomputed obsolete pdbs /data1/database/pdb_mmcif/obsolete.dat.
> I0418 15:18:45.851009 46912496321088 xla_bridge.py:226] Unable to initialize backend 'tpu_driver': NOT_FOUND: Unable to find driver in registry given worker: 
> 2024-04-18 15:18:45.870892: W external/org_tensorflow/tensorflow/stream_executor/platform/default/dso_loader.cc:64] Could not load dynamic library 'libcuda.so.1'; dlerror: libcuda.so.1: cannot open shared object file: No such file or directory; LD_LIBRARY_PATH: /.singularity.d/libs
> 2024-04-18 15:18:45.870957: W external/org_tensorflow/tensorflow/stream_executor/cuda/cuda_driver.cc:269] failed call to cuInit: UNKNOWN ERROR (303)
> I0418 15:18:45.871125 46912496321088 xla_bridge.py:226] Unable to initialize backend 'gpu': FAILED_PRECONDITION: No visible GPU devices.
> I0418 15:18:45.871670 46912496321088 xla_bridge.py:226] Unable to initialize backend 'tpu': INVALID_ARGUMENT: TpuPlatform is not available.
> W0418 15:18:45.871800 46912496321088 xla_bridge.py:230] No GPU/TPU found, falling back to CPU. (Set TF_CPP_MIN_LOG_LEVEL=0 and rerun for more info.)
> I0418 15:18:56.730648 46912496321088 run_alphafold.py:376] Have 5 models: ['model_1_pred_0', 'model_2_pred_0', 'model_3_pred_0', 'model_4_pred_0', 'model_5_pred_0']
> I0418 15:18:56.731153 46912496321088 run_alphafold.py:393] Using random seed 181191842569406945 for the data pipeline
> I0418 15:18:56.731739 46912496321088 run_alphafold.py:161] Predicting Brassica_napus_LOC106361359
> I0418 15:18:56.741892 46912496321088 jackhmmer.py:133] Launching subprocess "/opt/alphafold/alphafold-conda/bin/jackhmmer -o /dev/null -A /tmp/25812605.1.medium.q/tmpqziemejc/output.sto --noali --F1 0.0005 --F2 5e-05 --F3 5e-07 --incE 0.0001 -E 0.0001 --cpu 8 -N 1 Brassica_napus_LOC106361359.fa /data1/database/uniref90/uniref90.fasta"
> I0418 15:18:56.893545 46912496321088 utils.py:36] Started Jackhmmer (uniref90.fasta) query
> I0418 15:28:12.104611 46912496321088 utils.py:40] Finished Jackhmmer (uniref90.fasta) query in 555.211 seconds
> I0418 15:28:21.961827 46912496321088 jackhmmer.py:133] Launching subprocess "/opt/alphafold/alphafold-conda/bin/jackhmmer -o /dev/null -A /tmp/25812605.1.medium.q/tmpms1wxfkf/output.sto --noali --F1 0.0005 --F2 5e-05 --F3 5e-07 --incE 0.0001 -E 0.0001 --cpu 8 -N 1 Brassica_napus_LOC106361359.fa /data1/database/mgnify/mgy_clusters_2018_12.fa"
> I0418 15:28:22.091439 46912496321088 utils.py:36] Started Jackhmmer (mgy_clusters_2018_12.fa) query
> I0418 15:34:58.100716 46912496321088 utils.py:40] Finished Jackhmmer (mgy_clusters_2018_12.fa) query in 396.009 seconds
> I0418 15:35:16.404630 46912496321088 hhsearch.py:85] Launching subprocess "/opt/alphafold/alphafold-conda/bin/hhsearch -i /tmp/25812605.1.medium.q/tmpv2r73ccq/query.a3m -o /tmp/25812605.1.medium.q/tmpv2r73ccq/output.hhr -maxseq 1000000 -d /data1/database/pdb70/pdb70"
> I0418 15:35:16.683739 46912496321088 utils.py:36] Started HHsearch query
> I0418 15:41:07.724717 46912496321088 utils.py:40] Finished HHsearch query in 351.040 seconds
> I0418 15:41:13.168059 46912496321088 hhblits.py:128] Launching subprocess "/opt/alphafold/alphafold-conda/bin/hhblits -i Brassica_napus_LOC106361359.fa -cpu 4 -oa3m /tmp/25812605.1.medium.q/tmpe7d5g74o/output.a3m -o /dev/null -n 3 -e 0.001 -maxseq 1000000 -realign_max 100000 -maxfilt 100000 -min_prefilter_hits 1000 -d /data1/database/bfd/bfd_metaclust_clu_complete_id30_c90_final_seq.sorted_opt -d /data1/database/uniclust30/uniclust30_2018_08/uniclust30_2018_08"
> I0418 15:41:13.456561 46912496321088 utils.py:36] Started HHblits query
> I0418 17:49:16.971016 46912496321088 utils.py:40] Finished HHblits query in 7683.514 seconds
> E0418 17:49:16.979065 46912496321088 hhblits.py:138] HHblits failed. HHblits stderr begin:
> E0418 17:49:16.979293 46912496321088 hhblits.py:141] - 15:41:38.967 INFO: Searching 65983866 column state sequences.
> E0418 17:49:16.979351 46912496321088 hhblits.py:141] - 15:41:39.918 INFO: Searching 15161831 column state sequences.
> E0418 17:49:16.979410 46912496321088 hhblits.py:141] - 15:41:40.053 INFO: Brassica_napus_LOC106361359.fa is in A2M, A3M or FASTA format
> E0418 17:49:16.979455 46912496321088 hhblits.py:141] - 15:41:40.060 INFO: Iteration 1
> E0418 17:49:16.979498 46912496321088 hhblits.py:141] - 15:41:40.337 INFO: Prefiltering database
> E0418 17:49:16.979540 46912496321088 hhblits.py:141] - 17:23:47.203 INFO: HMMs passed 1st prefilter (gapless profile-profile alignment)  : 2394465
> E0418 17:49:16.979581 46912496321088 hhblits.py:141] - 17:48:28.171 INFO: HMMs passed 1st prefilter (gapless profile-profile alignment)  : 593904
> E0418 17:49:16.979622 46912496321088 hhblits.py:141] - 17:48:38.657 INFO: HMMs passed 2nd prefilter (gapped profile-profile alignment)   : 19095
> E0418 17:49:16.979663 46912496321088 hhblits.py:141] - 17:48:38.657 INFO: HMMs passed 2nd prefilter and not found in previous iterations : 19095
> E0418 17:49:16.979704 46912496321088 hhblits.py:141] - 17:48:38.657 INFO: Scoring 19095 HMMs using HMM-HMM Viterbi alignment
> E0418 17:49:16.979746 46912496321088 hhblits.py:141] - 17:48:38.771 INFO: Alternative alignment: 0
> E0418 17:49:16.979787 46912496321088 hhblits.py:142] HHblits stderr end
> Traceback (most recent call last):
>   File "/opt/alphafold/run_alphafold.py", line 422, in <module>
>     app.run(main)
>   File "/opt/alphafold/alphafold-conda/lib/python3.8/site-packages/absl/app.py", line 312, in run
>     _run_main(main, args)
>   File "/opt/alphafold/alphafold-conda/lib/python3.8/site-packages/absl/app.py", line 258, in _run_main
>     sys.exit(main(argv))
>   File "/opt/alphafold/run_alphafold.py", line 398, in main
>     predict_structure(
>   File "/opt/alphafold/run_alphafold.py", line 172, in predict_structure
>     feature_dict = data_pipeline.process(
>   File "/opt/alphafold/alphafold/data/pipeline.py", line 215, in process
>     hhblits_bfd_uniclust_result = run_msa_tool(
>   File "/opt/alphafold/alphafold/data/pipeline.py", line 96, in run_msa_tool
>     result = msa_runner.query(input_fasta_path)[0]
>   File "/opt/alphafold/alphafold/data/tools/hhblits.py", line 143, in query
>     raise RuntimeError('HHblits failed\nstdout:\n%s\n\nstderr:\n%s\n' % (
> RuntimeError: HHblits failed
> stdout:
> 
> 
> stderr:
> - 15:41:38.967 INFO: Searching 65983866 column state sequences.
> 
> - 15:41:39.918 INFO: Searching 15161831 column state sequences.
> 
> - 15:41:40.053 INFO: Brassica_napus_LOC106361359.fa is in A2M, A3M or FASTA format
> 
> - 15:41:40.060 INFO: Iteration 1
> 
> - 15:41:40.337 INFO: Prefiltering database
> 
> - 17:23:47.203 INFO: HMMs passed 1st prefilter (gapless profile-profile alignment)  : 2394465
> 
> - 17:48:28.171 INFO: HMMs passed 1st prefilter (gapless profile-profile alignment)  : 593904
> 
> - 17:48:38.657 INFO: HMMs passed 2nd prefilter (gapped profile-profile alignment)   : 19095
> 
> - 17:48:38.657 INFO: HMMs passed 2nd prefilter and not found in previous iterations : 19095
> 
> - 17:48:38.657 INFO: Scoring 19095 HMMs using HMM-HMM Viterbi alignment
> 
> - 17:48:38.771 INFO: Alternative alignment: 0

**Comment by yookuda** [comment, marked as answer]
メモリの割り当て量が足りないためにエラーになっています。
サンプルスクリプトの通りに
```
#$ -l s_vmem=160G
#$ -l mem_req=8G
#$ -pe def_slot 16
```
としてください。

HHblitsは実行にかなり大きな仮想メモリを要求するため、s_vmemを大きく取る必要があります。
この例では2.5TBの仮想メモリを割り当てている計算です。

**Reply by Shunsuke-Kanamori** [reply]
ご返信ありがとうございます。
そのようにメモリの割り当てをしてqsubすると以下のようなエラーが出てくるのですが、どうしたらジョブを投入できるでしょうか。

> Unable to run job: resource "s_vmem" and "mem_req" must be equivalent.
> Refer to : https://sc.ddbj.nig.ac.jp/ja/guide/usage-for-general-analysis-environment/uge
> Exiting.

**Reply by oogasawa** [reply]
mem_reqは物理メモリ量の指定、s_vmemは仮想メモリの指定で、この2つに違う値を指定してしまうと、その計算機上で走っているほかの人のジョブが影響を受けてしまいます。（ほかの計算機にスケジュールされるべきだったのにメモリの足りない計算機上で走り始めてしまうジョブが発生する可能性が出てくる。）

**Reply by oogasawa** [reply]
2.5TBの仮想メモリが必要な場合、mediumノードで走らせるか、計算ノードの占有の申請（有料）をするかになります。

mediumノードで走らせる場合は以下の通りとなります。

#$ -l s_vmem=160G
#$ -l mem_req=160G
#$ -pe def_slot 16
#$ -l medium

**Reply by Shunsuke-Kanamori** [reply]
ありがとうございます。
それでは、s_vmemとmem_reqの値を同じにした上で、メモリ割当量を上げてジョブを投入してみます。

**Reply by sc-fs** [reply]
JOB投入時にqsub_betaを利用されているのではないでしょうか。こちらのコマンドでは、実際に指定さえたs_vmemとmem_req
の差異をチェックしています。今回の指定のようにs_vmemとmem_reqの指定い差異がある場合は、qsubで実行してください。

**Reply by Shunsuke-Kanamori** [reply]
ありがとうございます。確かにqsub_betaで行なっていました。qsubに直して投入するように修正したいと思います。
